content/chapters/compute/lecture/slides.mdpp

```diff
@@ -0,0 +1,1088 @@
+
+![Thread Execution](media/thread-behaviour.png)
+
+----
+
+### Thread States
+
+* At most one RUNNING thread per core
+* Queues for BLOCKED and READY threads
+* [Quiz](./quiz/number-of-running-threads.md)
+
+![Thread State Diagram](media/thread-states.svg)
+
+A more detailed diagram is [here](#the-suspended-states)
+
+---
+
+### Context Switch - When?
+
+* Voluntary - initiated by the running thread:
+    * RUNNING thread performs a blocking action (eg: I/O call)
+    * The thread calls `yield()` / `SwitchToThread()`
+* Involuntary - initiated by the OS:
+    * RUNNING thread ends
+    * RUNNING thread's time slice expires
+    * A thread with higher priority is spawned
+
+----
+
+### Context Switch - How?
+
+![Context Switch and PCBs](media/context-switch-pcb.png)
+
+---
+
+### Types of Threads
+
+* CPU-intensive / CPU-bound = mostly CPU bursts:
+    * compute the digits of pi
+    * compilers
+    * machine learning
+* I/O-intensive / I/O-bound = mostly I/O bursts:
+    * servers
+    * databases
+    * text editors
+
+----
+
+### IO-bound vs CPU-bound
+
+* `demo/context-switch/cpu_bound.c`:
+```
+student@os:~$ cat /proc/$(pidof cpu_bound)/status
+[...]
+voluntary_ctxt_switches:        13
+nonvoluntary_ctxt_switches:     7
+```
+
+* `demo/context-switch/io_bound.c`:
+```
+student@os:~$ cat /proc/$(pidof io_bound)/status
+[...]
+voluntary_ctxt_switches:        3729
+nonvoluntary_ctxt_switches:     3
+```
+
+---
+
+### Context Switch - Overhead
+
+* Causes of overhead:
+    * running the scheduler itself
+    * after flushing the TLB of the evicted thread, the new thread starts with an empty TLB
+* Faster context switches between threads of the same process because their TLB is not flushed
+
+---
+
+### Scheduling - Ideals
+
+* **Interactivity:** decrease time spent in READY
+    * low waiting time
+    * also known as **fairness**
+* **Throughput:** finish threads as quickly as possible
+    * low turnaround time
+
+----
+
+### Scheduling - Reality
+
+* There is no free lunch
+* Tradeoff between interactivity and throughput
+
+----
+
+### Types of Schedulers
+
+* **Cooperative:** only allow voluntary context switches
+    * Advantage: Simple to implement
+    * Disadvantage: starvation - a thread might never yield its CPU core
+* **Preemptive:** also allows involuntary context switches
+    * Each process can run at most a certain _time slice_ / _quantum_
+    * When it expires, a context switch is triggered
+
+----
+
+### Scheduling Algorithms - The Bigger Picture
+
+* **Round-Robin**:
+    * Push each new thread to a queue
+    * Every time slice, dequeue one thread, run it, then reenqueue it
+    * Threads run in the order in which they're spawned
+* **Completely Fair Scheduler (CFS)**:
+    * Add new threads to a red-black tree sorted by total running time
+    * Every time slice, remove, run and add the thread that has run the least back to the tree
+    * Each thread runs approximately the same amount of time
```

Comment:
These lines are not properly displayed.

Reply:
I've split this slide into 2: one for Round-Robin and the other for the CFS. It's even clearer like this.